dist/dist_client.go

```diff
@@ -0,0 +1,500 @@
+			if username, password, err := config.GetAuthentication(sys, registry); err == nil {
+				creds = base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
+			}
+		}
+	}
+
+	r = OciRepo{ref: ref, authCreds: creds, client: client}
+
+	ping := func(scheme string) error {
+		u := url.URL{Scheme: scheme, Host: fmt.Sprintf("%s:%s", server, port), Path: fmt.Sprintf("/v2/")}
+		resp, err := client.Get(u.String())
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusUnauthorized {
+			return errors.Errorf("Error pinging registry %s:%s, response code %d (%s)",
+				server, port, resp.StatusCode, http.StatusText(resp.StatusCode))
+		}
+		return nil
+	}
+
+	scheme := "https"
+	err = ping(scheme)
+	if err != nil && insecureSkipVerify {
+		scheme = "http"
+		err = ping(scheme)
+	}
+	if err != nil {
+		return r, errors.Wrap(err, "unable to ping registry")
+	}
+
+	r.url = url.URL{Scheme: scheme, Host: fmt.Sprintf("%s:%s", server, port)}
+	return r, nil
+}
+
+func (o *OciRepo) GetManifest() ([]byte, *ispec.Manifest, error) {
+	name := o.ref.name
+	tag := o.ref.tag
+	m := &ispec.Manifest{}
+
+	var body []byte
+
+	manifestURI, err := url.Parse(fmt.Sprintf("/v2/%s/manifests/%s", name, tag))
+	if err != nil {
+		return body, m, errors.Wrapf(err, "couldn't parse manifest url for repo %s and image %s", name, tag)
+	}
+
+	manifestURI = o.url.ResolveReference(manifestURI)
+	req, err := http.NewRequest(http.MethodGet, manifestURI.String(), nil)
+
+	if err != nil {
+		return body, m, errors.Wrapf(err, "Couldn't create DELETE request for %v", manifestURI)
+	}
+
+	if o.authCreds != "" {
+		req.Header.Add("Authorization", "Basic "+o.authCreds)
+	}
+
+	req.Header.Add("Accept", "application/vnd.oci.image.manifest.v1+json")
+	resp, err := o.client.Do(req)
+
+	if err != nil {
+		return body, m, errors.Wrapf(err, "Error getting manifest %s %s from %v", name, tag, o.url)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return body, m, fmt.Errorf("bad return code %d getting manifest", resp.StatusCode)
+	}
+
+	body, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return body, m, errors.Wrapf(err, "Error reading response body for %s", tag)
+	}
+
+	err = json.Unmarshal(body, m)
+
+	if err != nil {
+		return body, m, errors.Wrap(err, "Failed decoding response")
+	}
+
+	return body, m, nil
+}
+
+func (o *OciRepo) RemoveManifest() error {
+	name := o.ref.name
+	tag := o.ref.tag
+	manifestURI, err := url.Parse(fmt.Sprintf("/v2/%s/manifests/%s", name, tag))
+
+	if err != nil {
+		return errors.Wrapf(err, "couldn't parse manifest url for repo %s and image %s", name, tag)
+	}
+
+	manifestURI = o.url.ResolveReference(manifestURI)
+	req, err := http.NewRequest(http.MethodDelete, manifestURI.String(), nil)
+
+	if err != nil {
+		return errors.Wrapf(err, "Couldn't create DELETE request for %v", manifestURI)
+	}
+
+	if o.authCreds != "" {
```

Comment:
(I can’t quite see the point of having code like this copy&pasted all over the place, when the original has this factored out to a single `dockerClient.makeRequest*` path.)